node/actors/bft/src/testonly/run.rs

```diff
@@ -71,77 +110,407 @@ impl Test {
+            }
+            anyhow::Ok(())
+        })
+        .await
+    })
+    .await
+}
+
+/// Run a set of nodes with a Twins network configuration.
+async fn run_nodes_twins(
+    ctx: &ctx::Ctx,
+    specs: &[Node],
+    splits: &PortSplitSchedule,
+) -> anyhow::Result<()> {
+    scope::run!(ctx, |ctx, s| async {
+        // All known network ports of a validator, so that we can tell if any of
+        // those addresses are in the same partition as the sender.
+        let mut validator_ports: HashMap<_, Vec<Port>> = HashMap::new();
+        // Outbox of consensus instances, paired with their network identity,
+        // so we can tell which partition they are in in the given view.
+        let mut recvs = vec![];
+        // Inbox of the consensus instances, indexed by their network identity,
+        // so that we can send to the one which is in the same partition as the sender.
+        let mut sends = HashMap::new();
+        // Blockstores of nodes, indexed by port; used to simulate the effect of gossip.
+        let mut stores = HashMap::new();
+        // Outbound gossip relationships to simulate the effects of block fetching.
+        let mut gossip_targets = HashMap::new();
+
+        for (i, spec) in specs.iter().enumerate() {
+            let (actor_pipe, dispatcher_pipe) = pipe::new();
+            let validator_key = spec.net.validator_key.as_ref().unwrap().public();
+            let port = spec.net.server_addr.port();
+
+            validator_ports.entry(validator_key).or_default().push(port);
+
+            sends.insert(port, actor_pipe.send);
+            recvs.push((port, actor_pipe.recv));
+            stores.insert(port, spec.block_store.clone());
+            gossip_targets.insert(
+                port,
+                spec.net
+                    .gossip
+                    .static_outbound
+                    .values()
+                    .map(|host| {
+                        let addr: std::net::SocketAddr = host.0.parse().expect("valid address");
+                        addr.port()
+                    })
+                    .collect(),
+            );
+
+            // Run consensus; the dispatcher pipe is its network connection, which means we can use the actor pipe to:
+            // * send Output messages from other actors to this consensus instance
+            // * receive Input messages sent by this consensus to the other actors
+            s.spawn(
+                async {
+                    let mut network_pipe = dispatcher_pipe;
+                    spec.run(ctx, &mut network_pipe).await
                 }
+                .instrument(tracing::info_span!("node", i)),
+            );
+        }
+        // Taking these references is necessary for the `scope::run!` environment lifetime rules to compile
+        // with `async move`, which in turn is necessary otherwise it the spawned process could not borrow `port`.
+        // Potentially `ctx::NoCopy` could be used with `port`.
```

Comment:
nit: NoCopy would be preferrable, as `scope::run` is intended to be supplied with `async`, not `async move`.